File: frontend-2024-grupo2/src/app/data/models/Show.ts
import { PlaceOfShow, PlaceOfShowJSON } from './PlaceOfShow'
import { StatusShow, StatusShowJSON } from './StatusShow'
import { FunctionShow, FunctionShowJSON } from './FunctionShow'
import { UserFriendJSON } from './User'

export type ShowJSON = {
  id: number
  bandName: string
  placeOfShow: PlaceOfShowJSON
  nameOfRecital: string
  img: string
  status: StatusShowJSON
  costBand: number
  functionShow: FunctionShowJSON[]
  friendWithShow: UserFriendJSON[]
  score: number
  amountOfComments: number
  maxPrice: number
  minPrice: number
  userCommented: boolean
  ticketPriceForUser: number
}

export type FilterShowJSON = {
  bandName: string
  location: string
  byFriends: boolean
}

export class Show {
  constructor(
    public id: number,
    public bandName: string,
    public placeOfShow: PlaceOfShow,
    public nameOfRecital: string,
    public img: string,
    public status: StatusShow,
    public costBand: number,
    public functionShow: FunctionShow[],
    public friends: UserFriendJSON[],
    public score: number,
    public amountOfComments: number,
    public maxPrice: number,
    public minPrice: number,
    public userCommented: boolean,
    public ticketPriceForUser: number
  ) {}

  static fromJSON(showJSON: ShowJSON): Show {
    return Object.assign(
      new Show(
        showJSON.id,
        showJSON.bandName,
        PlaceOfShow.fromJSON(showJSON.placeOfShow),
        showJSON.nameOfRecital,
        showJSON.img,
        StatusShow.fromJSON(showJSON.status),
        showJSON.costBand,
        showJSON.functionShow.map((functionJSON) =>
          FunctionShow.fromJSON(functionJSON)
        ),
        showJSON.friendWithShow,
        showJSON.score,
        showJSON.amountOfComments,
        showJSON.maxPrice,
        showJSON.minPrice,
        showJSON.userCommented,
        showJSON.ticketPriceForUser
      )
    )
  }

  toJSON(): ShowJSON {
    return {
      id: this.id,
      bandName: this.bandName,
      placeOfShow: this.placeOfShow.toJSON(),
      nameOfRecital: this.nameOfRecital,
      img: this.img,
      status: this.status.toJSON(),
      costBand: this.costBand,
      functionShow: this.functionShow.map((functionShow) =>
        functionShow.toJSON()
      ),
      friendWithShow: this.friends,
      score: this.score,
      amountOfComments: this.amountOfComments,
      maxPrice: this.maxPrice,
      minPrice: this.minPrice,
      userCommented: this.userCommented,
      ticketPriceForUser: this.ticketPriceForUser
    }
  }
}

File: backend-2024-grupo2/src/main/kotlin/ar/edu/unsam/phm/backendgrupo2/domain/Show.kt
package ar.edu.unsam.phm.backendgrupo2.domain

import ar.edu.unsam.phm.backendgrupo2.dto.ShowDTO
import jakarta.persistence.*
import java.math.BigDecimal

@Entity
@Table(name = "show")
class Show {
	
	companion object {
		
		fun fromJSON(showDTO: ShowDTO): Show =
			Show().apply {
				this.id = showDTO.id
				this.bandName = showDTO.bandName
				this.placeOfShow = PlaceOfShow.fromJSON(showDTO.placeOfShow)
				this.nameOfRecital = showDTO.nameOfRecital
				this.costBand = showDTO.costBand.toBigDecimal()
				this.status = StatusShow.fromJSON(showDTO.status)
				this.image = showDTO.image
				this.functionShowList = showDTO.functionShow.map { FunctionShow.fromJSON(it) }.toMutableList()
			}
	}

	@Id
	@GeneratedValue
	var id: Int = 0

	@Column(length = 40)
	var bandName: String = ""

	@Column(length = 40)
	var nameOfRecital: String = ""
	
	@Column(length = 40)
	var image: String = ""

	@Column
	var costBand: BigDecimal = 0.toBigDecimal()

	@ManyToOne(fetch = FetchType.EAGER)
	var status: StatusShow? = null
	
	@ManyToOne(fetch = FetchType.EAGER)
	var placeOfShow: PlaceOfShow? = null

	@ManyToMany(fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
	var functionShowList: MutableList<FunctionShow> = mutableListOf()
	
	@OneToMany(fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
	var commentList: MutableList<Comment> = mutableListOf()

	fun amountPeopleWaiting(): Int = 0

	fun amountFunctionSoldOut(): Int =
		this.functionShowList.count { it.isSoldOut() }

	fun totalSales(): BigDecimal =
		this.functionShowList.sumOf { it.totalSales() }
	
	fun score(): Double =
		if(this.commentList.size <= 0) 0.toDouble()
		else (this.commentList.sumOf { it.score } / this.commentList.size).toDouble()
	
	fun totalAmountOfTickets(): Int =
		this.functionShowList.sumOf { it.amountOfTicketsSold() }

	fun totalAmountOfTicketsPerSeatType(seatType: SeatType): Int =
		this.functionShowList.sumOf { it.ticketsBySeatType(seatType) }

	fun cost(): BigDecimal =
		this.placeOfShow!!.baseCost() + this.costBand

	fun minTicketPrice(): BigDecimal =
		if(this.functionShowList.isNotEmpty()) this.functionShowList.minOf { it.minTicketPrice() }
		else 0.toBigDecimal()

	fun maxTicketPrice(): BigDecimal =
		if(this.functionShowList.isNotEmpty()) this.functionShowList.maxOf { it.maxTicketPrice() }
		else 0.toBigDecimal()

	fun amountOfComments(): Int =
		this.commentList.size

	fun addCommentsToShow(comment: List<Comment>) {
		this.commentList.addAll(comment)
	}

	fun addComment(comment: Comment) {
		this.commentList.add(comment)
	}

	fun getComments(): List<Comment> =
		this.commentList

	fun addFunctionShow(functionShow: FunctionShow) {
		this.functionShowList.add(functionShow)
	}

	fun addFunctionsToShow(newFunctionShowList: List<FunctionShow>) {
		this.functionShowList.addAll(newFunctionShowList)
	}
	
	fun getCommentFromUser(user: User): List<Comment> =
		this.commentList.filter { it.owner?.id == user.id }

	fun userFriends(user: User): List<User> =
		user.getFriends().filter { it.getShows().contains(this) }

	fun userCommented(user: User): Boolean =
		this.commentList.any { it.owner?.id == user.id }

	fun hasComment(comment: Comment): Boolean =
		this.commentList.contains(comment)

	fun removeComment(comment: Comment) {
		this.commentList.remove(comment)
	}
}